bin/i18n/i18n_script_utils.rb

```diff
@@ -135,30 +135,47 @@ def self.get_level_url_key(script, level)
   def self.get_level_from_url(url)
     # memoize to reduce repeated database interactions
     @levels_by_url ||= Hash.new do |hash, new_url|
-      url_regex = %r{https://studio.code.org/s/(?<script_name>[A-Za-z0-9\s\-_]+)/stage/(?<stage_pos>[0-9]+)/(?<level_info>.+)}
-      matches = new_url.match(url_regex)
-
-      hash[new_url] =
-        if matches.nil?
-          project_url_regex = %r{https://studio.code.org/p/(?<project_name>[A-Za-z0-9\s\-_]+)}
-          project_matches = new_url.match(project_url_regex)
-          if project_matches.nil?
-            STDERR.puts "could not find level for url: #{new_url}"
-            nil
+      route_params = Rails.application.routes.recognize_path(new_url)
```

Comment:
this seems like a much more solid approach! this will also hopefully be much easier to update when we change how course/script/level urls work next year. Heads up that this is scheduled to happen in Q1... sorry, we are trying to get better about sharing this kind of info out.

I see some code in here to handle stage extras, lockable lessons, and flappy routes, which is great! Going through the rolodex of other special level types, I will just note that I do not see any code in here to handle:
* level-swapping (level shown depends on signed-in user, if more than one variant exists)
* `LevelGroup` pages: https://github.com/code-dot-org/code-dot-org/blob/dfb989b0d0c746b40d98a6f6d9b682c30dbb4aae/dashboard/config/routes.rb#L284-L285
* `BubbleChoice` sublevels: https://github.com/code-dot-org/code-dot-org/blob/dfb989b0d0c746b40d98a6f6d9b682c30dbb4aae/dashboard/config/routes.rb#L286-L287

I haven't dug into the purpose of this code, though, so I can't say which of these are important or not.

Reply:
> when we change how course/script/level urls work next year

oh, good to know! That shouldn't be too disruptive for this specific code, but will _definitely_ require a significant undertaking on the other end. We use the level URLs as keys for all level-specific i18n strings, so that change will likely require a sizeable Crowdin migration. We should coordinate as the time gets closer.

> I do not see any code in here to handle:

None of those routes are currently relevant for the data this operates on, but I'll make sure to keep an eye out for any changes in this area that could affect that.

Reply:
> > when we change how course/script/level urls work next year
> 
> oh, good to know! That shouldn't be too disruptive for this specific code, but will _definitely_ require a significant undertaking on the other end. We use the level URLs as keys for all level-specific i18n strings, so that change will likely require a sizeable Crowdin migration. We should coordinate as the time gets closer.

Good to know too :-) We are considering nailing down the new URL plan in January and shipping it by April. Does that sound like a reasonable timeframe for taking i18n considerations into account? We think a cleanup of our curriculum URLs will make everyone's lives easier in the long run, but want to be sensitive to the work this puts on other teams.